core/src/apps/base.py

```diff
@@ -457,7 +490,12 @@ def boot() -> None:
         workflow_handlers.register(msg_type, handler)  # type: ignore [cannot be assigned to type]
 
     reload_settings_from_storage()
-    if config.is_unlocked():
+    if (
+        storage_cache.get(storage_cache.APP_RECOVERY_REPEATED_BACKUP_UNLOCKED)
+        == b"\x01"
+    ):
+        wire.find_handler = _get_backup_handler
+    elif config.is_unlocked():
         wire.find_handler = workflow_handlers.find_registered_handler
     else:
-        wire.find_handler = get_pinlocked_handler
+        wire.find_handler = _get_pinlocked_handler
```

Comment:
Very good point. The question of course is - if we are to zoom out and assume things will evolve in unexpected ways - do we know for sure which list is a subset of which? Maybe things will not stay as they are now forever, and a simple change in the future which would not get such an in-depth review will slip through... (no pun intended)

So maybe

> [...] all calls to find_registered_handler() be replaced with a dispatch function which looks-up [...]

a variation on this idea, but somewhat more general - a dispatch function which looks for both restricted cases (repeated backup *and* locked screen) and picks the intersection of the two sets if necessary (possibly to be extended to N sets in the future).

But if we do that ... I am thinking - should we even ever overwrite `wire.find_handler` at all? Or would it always be this dispatch function checking the restricted cases anyway, so we would get rid of the need to swap the `find_handler` when state changes?

Reply:
i'll note that this is completely irrelevant here because when autolock triggers, it will shut down the backup screen and this will clear the cache entry so only the pinlocked handler will stay active

but with that said we also have the option of chaining the handlers.
super rough draft:
```
def get_pinlocked_handler():
    prev_handler = wire.find_handler
    def handler(msg):
          if i_am_handling_this(msg):
               _handle_this(msg)
          else:
               return prev_handler(msg)
```

Reply:
> when autolock triggers, it will shut down the backup screen and this will clear the cache entry

Hmm, I didn't realize that. Even now that it has been stated, I am not sure how that works. Anyway, I think it's not something that we should rely on, because it seems like that could easily change, e.g. because the product team might not be happy with this behavior or some completely independent code change. Designing these sort of things well from the start can save some expensive refactoring in the future.

Reply:
https://github.com/trezor/trezor-firmware/pull/3743/commits/0ee80abf9193df39b298a80a9d5813979503ec3e